### PR TITLE
[17.0][MIG] contract_variable_quantity: Migration to version 17.0

### PR DESCRIPTION
**Pull request by juanjosesegui-tecnativa**
@Tecnativa TT50113

### CODE REVIEW

**Comment by juanjosesegui-tecnativa** [comment]
ping @pedrobaeza

File: contract_variable_quantity/readme/USAGE.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+To use this module, you need to:
+
+1.  Go to Invoicing \> Sales \> Contracts and select or create a new
```
Comment (carolinafernandez-tecnativa) [inline contract_variable_quantity/readme/USAGE.md]: This should be: Go to Invoicing \> Customers \> Customers Contracts

File: contract_variable_quantity/views/abstract_contract_line.xml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='quantity']" position="before">
+                <field name="qty_type" />
+            </xpath>
+            <xpath expr="//field[@name='quantity']" position="after">
+                <field
+                    name="qty_formula_id"
+                    required="('qty_type', '=', 'variable')"
+                    invisible="('qty_type', '!=', 'variable')"
```
Comment (carolinafernandez-tecnativa) [inline contract_variable_quantity/views/abstract_contract_line.xml]: required="qty_type == 'variable'"
invisible="qty_type != 'variable'"

File: contract_variable_quantity/views/contract_line_views.xml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            name="name"
+        >contract.line tree view (in contract) - Add variable qty fields</field>
+        <field name="model">contract.line</field>
+        <field name="inherit_id" ref="contract.contract_line_tree_view" />
+        <field name="arch" type="xml">
+            <field name="quantity" position="before">
+                <field name="qty_type" />
+                <field
+                    name="qty_formula_id"
+                    invisible="[('qty_type', '!=', 'variable')]"
```
Comment (carolinafernandez-tecnativa) [inline contract_variable_quantity/views/contract_line_views.xml]: column_invisible="qty_type != 'variable'"
Please check https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0

File: contract_variable_quantity/views/contract_template.xml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+                <field name="qty_type" />
+            </xpath>
+            <xpath
+                expr="//field[@name='contract_line_ids']//field[@name='quantity']"
+                position="after"
+            >
+                <field
+                    name="qty_formula_id"
+                    required="[('qty_type', '=', 'variable')]"
+                    invisible="[('qty_type', '!=', 'variable')]"
```
Comment (carolinafernandez-tecnativa) [inline contract_variable_quantity/views/contract_template.xml]: same as comments before.

**Review by carolinafernandez-tecnativa — COMMENTED** [review]
1- Please squash administrative commits.
2- When creating a new customer contract i got this error message

image

Please check this contract configuration that is not working

image

Attributes invisible,required are not well defined so im not able for example as introducing qty formula

Thanks

File: contract_variable_quantity/views/contract.xml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+                position="attributes"
+            >
+                <attribute name="invisible">"qty_type != 'fixed'"</attribute>
+            </xpath>
+            <xpath
+                expr="//field[@name='contract_line_fixed_ids']/tree/field[@name='quantity']"
+                position="before"
+            >
+                <field name="qty_type" />
+                <field name="qty_formula_id" invisible="qty_type != 'variable'" />
```
Comment (carolinafernandez-tecnativa) [inline contract_variable_quantity/views/contract.xml:28]: in tree view, if you want to make a column invisible you have to put column_invisible, please check this for the rest of fields in tree view

File: contract_variable_quantity/views/contract_line_formula.xml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        <field name="arch" type="xml">
+            <form>
+                <sheet>
+                    <div class="oe_title">
+                        <h1>
+                            <field name="name" placeholder="Name" />
+                        </h1>
+                    </div>
+                    <group string="Code">
+                        <div style="margin-top: 4px;">
```
Comment (carolinafernandez-tecnativa) [inline contract_variable_quantity/views/contract_line_formula.xml]: This is displaying weird, please review it, something must be changed, review with others module version

File: contract_variable_quantity/views/contract_line_views.xml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        >contract.line tree view (in contract) - Add variable qty fields</field>
+        <field name="model">contract.line</field>
+        <field name="inherit_id" ref="contract.contract_line_tree_view" />
+        <field name="arch" type="xml">
+            <field name="quantity" position="before">
+                <field name="qty_type" />
+                <field name="qty_formula_id" invisible="qty_type != 'variable'" />
+            </field>
+            <field name="quantity" position="attributes">
+                <attribute name="invisible">"qty_type != 'fixed'"</attribute>
```
Comment (carolinafernandez-tecnativa) [inline contract_variable_quantity/views/contract_line_views.xml:17]: same column_invisible

**Review by carolinafernandez-tecnativa — COMMENTED** [review]
Check pressing Send by email button in a customer contact, it is displaying error message, thanks

File: contract_variable_quantity/views/contract.xml (hunk 5 above)
Comment (juanjosesegui-tecnativa) [reply contract_variable_quantity/views/contract.xml:28]: column_invisible cannot be evaluated on a per-record basis. It affects the entire column, only accepts true or false values, and does not consider subtree values.

**Comment by juanjosesegui-tecnativa** [comment]
All changes have been made.

@carolinafernandez-tecnativa

File: contract_variable_quantity/views/contract_line_views.xml (hunk 3 above)
Comment (pedrobaeza) [reply contract_variable_quantity/views/contract_line_views.xml]: The migration guide states:

```
On a tree view, if you have a field with invisible="1", you should change it to column_invisible="1", or now the column will be shown, but empty.
```

but this is making the visibility conditionally according some record values. That continues the same.

Comment (pedrobaeza) [reply contract_variable_quantity/views/contract_line_views.xml]: I was about to add some comment in the migration guide, but this change comes from `attrs`, so nothing more to clarify, as it's totally different.

**Review by pedrobaeza — APPROVED** [review]
Code and functional review.

**Review by carolinafernandez-tecnativa — APPROVED** [review]
Functional review LGTM, Thanks :)

File: contract_variable_quantity/views/contract_template.xml [hunk 8]
```diff
@@ -17,16 +17,16 @@
...
             </xpath>
             <xpath
                 expr="//field[@name='contract_line_ids']//field[@name='quantity']"
                 position="attributes"
             >
-                <attribute
-                    name="attrs"
-                >{'required': [('qty_type', '=', 'fixed')], 'invisible': [('qty_type', '!=', 'fixed')]}</attribute>
+                <attribute name="required">"qty_type == 'fixed'"</attribute>
+                <attribute name="invisible">"qty_type != 'fixed'"</attribute>
```
Comment (javierjcf) [inline contract_variable_quantity/views/contract_template.xml:29]: Careful here! The expression must not be a string. Refer to #1242 for the fix.

**Pull request merged**